# Exercice 1 : À la découverte des Variables (à faire dans la console)

In [1]:
# Q1.
x = 5

Q2. Non on ne peut pas écrire l’égalité d’affectation dans le sens qu’on veut.

In [2]:
5 = x

SyntaxError: cannot assign to literal (Temp/ipykernel_14944/1182432656.py, line 1)

In [3]:
# Q3. 
y = 7 * x; print(y)

35


In [4]:
x = 2; print(y)

35


In [5]:
# Q4.
help(id)

Help on built-in function id in module builtins:

id(obj, /)
    Return the identity of an object.
    
    This is guaranteed to be unique among simultaneously existing objects.
    (CPython uses the object's memory address.)



In [6]:
help(type)

Help on class type in module builtins:

class type(object)
 |  type(object_or_name, bases, dict)
 |  type(object) -> the object's type
 |  type(name, bases, dict) -> a new type
 |  
 |  Methods defined here:
 |  
 |  __call__(self, /, *args, **kwargs)
 |      Call self as a function.
 |  
 |  __delattr__(self, name, /)
 |      Implement delattr(self, name).
 |  
 |  __dir__(self, /)
 |      Specialized __dir__ implementation for types.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __init__(self, /, *args, **kwargs)
 |      Initialize self.  See help(type(self)) for accurate signature.
 |  
 |  __instancecheck__(self, instance, /)
 |      Check if an object is an instance.
 |  
 |  __repr__(self, /)
 |      Return repr(self).
 |  
 |  __setattr__(self, name, value, /)
 |      Implement setattr(self, name, value).
 |  
 |  __sizeof__(self, /)
 |      Return memory consumption of the type object.
 |  
 |  __subclasscheck__(self, subclass, /)
 |     

In [7]:
type(x)

int

In [8]:
id(x)

1830426863952

In [9]:
#Q5.
x =x + 1; id(x)

1830426863984

In [10]:
# Q6.
x = 11
x +=1; print(x)
x -=1; print(x)
x //=2; print(x)
x %=2; print(x)
x /=2; print(x)
x *=22; print(x)

12
11
5
1
0.5
11.0


In [11]:
#Q7.
x, y = 7, 9
x, y = y, x
print(x, y)

9 7


# Exercice 2 : Premiers affichages (à faire dans la console)

In [12]:
# Q1
Bonjour = 0

In [13]:
# Q2
print(Bonjour)

0


In [14]:
print("Bonjour") # ou print('Bonjour')

Bonjour


In [15]:
# Q3.
type(Bonjour)

int

In [16]:
type('Bonjour')

str

In [17]:
#Q4
print('Aujourd\'hui et Il dit: "Bonjour"') # \' est un carctère (') hors syntaxe python.

Aujourd'hui et Il dit: "Bonjour"


In [18]:
print("\"Aujourd’hui\"") # \" est un carctère (") hors syntaxe python.

"Aujourd’hui"


In [19]:
# Q5
print("Hello World!")

Hello World!


In [20]:
print("Hello\nWorld!") # \n  est un caractère de retour à la ligne

Hello
World!


In [21]:
# Q6 
x, y, z = 1, 2, 3
print(x, y, z, sep = ";")

1;2;3


In [22]:
print(x, y, z, sep = "et", end = '.')

1et2et3.

In [23]:
print(x, y, z, sep = "\n")

1
2
3


In [24]:
# Q7.
print("x={}, y={}, z={}".format(x, y, z))

x=1, y=2, z=3


# Exercice 4 : Quiz multiplications

In [26]:
# Quiz multiplication
import random as rd
## Q1)
a = rd.randint(1, 12)
## Q2)
b = rd.randint(1, 12)
## Q3)
print("combien vaut le produit {} * {} ?".format(a, b))
## Q4)
p = int(input())
## Q5)
if p == a*b:
    print("Bravo!")
else:
    print("Perdu! La bonne réponse était ",a*b)

combien vaut le produit 9 * 12 ?
13
Perdu! La bonne réponse était  108


# Exercice 5 : Structures itératives
Le programme affiche un triangle d'étoiles.

In [27]:
n = int(input("Entrez le nombre de ligne : "))
for i in range(n) :
    for j in range(i+1) :
        print(" ∗ ", end = "")
    print("\r")

Entrez le nombre de ligne : 6
 ∗ 
 ∗  ∗ 
 ∗  ∗  ∗ 
 ∗  ∗  ∗  ∗ 
 ∗  ∗  ∗  ∗  ∗ 
 ∗  ∗  ∗  ∗  ∗  ∗ 


# Exercice 6 : Aire du rectangle par rapport au cercle

In [28]:
from math import pi
r = 10.6
a = 1.3    # un côté du rectangle
aire_cercle = pi*r**2
b = 0      # valeur de départ choisie pour l'autre côté du rectangle
while a*b < aire_cercle:
    b += 1
b -= 1     # doit annuler la dernière mise à jour pour obtenir la bonne valeur
print("La plus grande valeur possible de b : ", b)

La plus grande valeur possible de b :  271


# Exercice 7 : Les fonctions

In [29]:
import math
def ecart(a, b, epsilon = 0.1) :
    d = math.fabs(a - b)
    if (d < epsilon) :
        d = 0
    return d

In [30]:
#Q1
print(ecart(a=12.2, b=11.9, epsilon = 1))

0


In [31]:
print(ecart(a=12.2, b=11.9))

0.29999999999999893


In [32]:
e= ecart(a=12.2, b=11.9) ;print(e)

0.29999999999999893


In [33]:
# Q2.
import math
def ecart(a, b, epsilon = 0.1) :
    d = math.fabs(a - b)
    if (d < epsilon) :
        d = 0
    print(d)

In [34]:
e1=ecart(a=12.2, b=11.9) ;print(e1)

0.29999999999999893
None


`0.29999999999999893` provient de l'affichage avec `print` dans `ecart()`.

`print(e1)` affiche la valeur `None` car la fonction `ecart()` ne retourne rien.


In [35]:
ecart(a=12.2, b=11.9)

0.29999999999999893


# Exercice 8 : date plus un jour

## Q1. Déterminer si une année est bissextile ou non?

In [36]:

def bissextile(a):
    return (a % 4 == 0 and a %100 != 0) or a % 400 ==0

In [37]:
bissextile(2021)

False

In [38]:
bissextile(2020)

True

## Q2. Donner le triplet (jour, mois et année) et affiche la date suivante.
* Les mois numéros 1, 3, 5, 7, 8, 10 et 12 ont 31 jours.
* Ceux de numéros 4, 6, 9 et 11 en ont 30.
* Le mois numéro 2 a 29 jours si l’année est bissextile, 28 dans le cas contraire.

### Une méthode

In [39]:
def plus_un_jour(j, m, a):
    if m in [1, 3, 5, 7, 8, 10] :
        if j == 31:
            j, m, a = 1, m+1,a
        else:
            j, m, a = j+1, m, a
    elif m == 12 :
        if j == 31:
            j, m, a = 1, 1, a+1
        else:
            j, m, a = j+1, m, a
    elif m in [4, 6, 9, 11]:
        if j == 30:
            j, m, a = 1, m+1,a
        else:
            j, m, a = j+1, m, a
    
    elif (j==28 and m==2 and not bissextile(a)) or (j==29 and m==2 and bissextile(a)):
        j, m, a = 1, m+1,a
    # La date à retouner par la fonction doit être valide!
    elif (m == 2 and not bissextile(a)and j>=29)or (m == 2 and bissextile(a)and j>29):
        return "Date invalide!"
    else :
        j, m, a = j+1, m, a
    # La date à retouner par la fonction doit être valide!
    if (1<= j <= 31 and 1 <= m <=12 and a >=1):
        return j, m, a
    else:
        return "Date invalide!"

In [40]:
 plus_un_jour(31, 6, 2015)

'Date invalide!'

In [41]:
 plus_un_jour(28, 2, 2020)

(29, 2, 2020)

In [42]:
plus_un_jour(30, 4, 2019)

(1, 5, 2019)

In [43]:
plus_un_jour(30, 3, 2020)

(31, 3, 2020)

In [44]:
plus_un_jour(29, 2, 2021)

'Date invalide!'

In [45]:
plus_un_jour(29, 2, 2020)

(1, 3, 2020)

### Une autre méthode
Une solution plus simple est possible en définissant une fonction intermédiaire qui renvoie le nombre de jours pour un mois et une année donnés.

In [46]:
def nbrJours(m, a):
    if m == 2:
        if bissextile(a):
            nbr = 29
        else:
            nbr = 28
    elif m <= 7:
        if m % 2 == 1:
            nbr = 31
        else:
            nbr = 30
    else:
        if m % 2 == 0:
            nbr = 31
        else:
            nbr = 30
    return nbr

In [47]:
nbrJours(2, 2021)

28

In [48]:
def Plus_Un_Jour(j, m, a):
    if j < nbrJours(m, a): # si j avant le dernier jour
        j += 1
    else: # si j est le dernier jour
        j = 1 # le lendemain est le premier jour
        if m < 12: # du nouveau mois
            m += 1
        else: # de la nouvelle année
            m = 1
            a += 1
    return j, m, a

In [49]:
Plus_Un_Jour(31, 12, 2015)

(1, 1, 2016)

In [50]:
Plus_Un_Jour(28, 2, 2020)

(29, 2, 2020)

In [51]:
Plus_Un_Jour(28, 2, 2021)

(1, 3, 2021)

# Exercice 9 : Nombres entiers

In [52]:
# Q1
def inf(n):
    i = 0
    while i < n:
        print(i)
    i+=1
# Q2
def inf_pair(n):
    i = 0
    while i < n:
        if i%2==0:
            print(i)
        i+=1
# Q3          
def div(n):
    i = 1
    while i <=n:
        if n%i ==0 :
            print(i)
        i+=1
# Q4
def ndiv(n):
    ndiv = 0
    i = 1
    while i <=n:
        if n%i ==0 : # si i divise n, on ajoute 1 à ndiv
            ndiv+=1
        i+=1
    return ndiv
# Q5
def somme_div(n):
    s = 0
    i = 1
    while i <=n:
        if n%i ==0 : # si i divise n, on l'joute à s
            s+=i
        i+=1
    return s
# Q6 # Q7
def MaxMin(n):
    i = 0
    Max=1
    Min = Max
    posMin, posMax = 0, 0
    while i<=n:
        x = int(input("Donnez un entier :"))
        if x>Max:
            Max = x
            posMax = i
        elif x<=Min:
            Min = x
            posMin = i
        i+=1
    return (Max, posMax), (Min, posMin)
# Q8.
3def MaxMin2():
    i = 0
    Max=1
    Min = Max
    posMin, posMax = 0, 0
    while True:
        x = int(input("Donnez un entier:"))
        if x>Max:
            Max = x
            posMax = i
        elif x<=Min:
            Min = x
            posMin = i
        if x == 0:
            break
        i+=1
    return (Max, posMax), (Min, posMin)

In [53]:
MaxMin(5)

Donnez un entier :10
Donnez un entier :11
Donnez un entier :22
Donnez un entier :0
Donnez un entier :-12
Donnez un entier :01


((22, 2), (-12, 4))

In [54]:
MaxMin2()

Donnez un entier:3
Donnez un entier:12
Donnez un entier:0


((12, 1), (0, 2))

# Exercice 10 : Calculette

In [55]:
#Q1
def saisi():
    while True:
        try:
            a = int(input("Donnez un entier a : "))
            b = int(input("Donnez un entier b : "))
            if a > 0 and b > 0:
                break
        except:
            print("Erreur")
    return (a, b)

In [60]:
a, b = saisi()

Donnez un entier a : 12
Donnez un entier b : 0
Donnez un entier a : aa
Erreur
Donnez un entier a : 12
Donnez un entier b : 3


In [61]:
#Q2
def somme(x, y):
    s = x
    for i in range(y):
        s+=1
    return s

In [62]:
somme(a, b)

15

In [63]:
#Q3
def difference(x, y):
    d = x
    for i in range(y):
        d-=1
    return d

In [64]:
difference(a, b)

9

In [65]:
#Q4
def produit(x, y):
    p = 0
    for i in range(y):
        p = somme(p, x)
    return p

In [66]:
produit(a, b)

36

In [67]:
#Q5
def puissance(x, y):
    e = 1
    for i in range(y):
        e = produit(e, x)
    return e

In [68]:
puissance(a, b)

1728

In [69]:
puissance(a,b) == a**b

True

In [73]:
#Q6
def div_mod(x, y):
    q = 0
    while produit(q, y) <= x:
        q +=1
    q-=1
    r = difference(x, produit(q, y))
    return(q, r)

In [74]:
div_mod(a, b)

(4, 0)

In [75]:
div_mod(a, b) == divmod(a, b)

True

# Exercice 11 : Heure valide

In [77]:
# Q1.
def saisi(a, b):
    while True:
        try:
            n = int(input("Donner n dans [a, b]"))
            if a<=n<=b:
                break
        except:
            print("Erreur, n doit être un entier dans [a, b]")
    return n

In [78]:
#Q2
def saisi_heure():
    print("Heure : ")
    hh = saisi(0, 23)
    print("Minute : ")
    mm = saisi(0, 59)
    print("Seconde : ")
    ss = saisi(0, 59)
    return (hh, mm, ss)
    
    

In [79]:
saisi_heure()

Heure : 
Donner n dans [a, b]12
Minute : 
Donner n dans [a, b]30
Seconde : 
Donner n dans [a, b]1


(12, 30, 1)

In [80]:
#Q3.
def H2S(h, m, s):
    return h*3600 + m*60 + s
def S2H(s):
    h = s//3600
    r = s%3600
    m = r//60
    s = r%60
    return (h, m, s)

In [81]:
def duree_voyage(h1, m1, s1, h2, m2, s2):
    s = H2S(h2, m2, s2) - H2S(h1, m1, s1)
    return S2H(s)


In [86]:
#Q4.
print("Heure du départ\n----------")
h1, m1, s1 = saisi_heure()
print("départ à {:02d}:{:02d}:{:02d}".format(h1,m1,s1)) # formate la chaîne d'entier à deux chiffres
print("Heure d'arrivée\n---------- ")
h2, m2, s2 = saisi_heure()
print("Arrivée à {:02d}:{:02d}:{:02d}".format(h2,m2,s2))
print("Durée de voyage\n----------")
h, m, s = duree_voyage(h1, m1, s1, h2, m2, s2)
print("{:02d}:{:02d}:{:02d}".format(h,m,s))

Heure du départ
----------
Heure : 
Donner n dans [a, b]12
Minute : 
Donner n dans [a, b]10
Seconde : 
Donner n dans [a, b]1
départ à 12:10:01
Heure d'arrivée
---------- 
Heure : 
Donner n dans [a, b]13
Minute : 
Donner n dans [a, b]30
Seconde : 
Donner n dans [a, b]0
Arrivée à 13:30:00
Durée de voyage
----------
01:19:59
